finesse/gui/hardware_set/device_view.py

```diff
@@ -0,0 +1,317 @@
+    QHBoxLayout,
+    QPushButton,
+    QSizePolicy,
+    QVBoxLayout,
+    QWidget,
+)
+
+from finesse.device_info import DeviceBaseTypeInfo, DeviceInstanceRef, DeviceTypeInfo
+from finesse.gui.error_message import show_error_message
+from finesse.gui.hardware_set.device_connection import close_device, open_device
+from finesse.gui.hardware_set.hardware_set import OpenDeviceArgs
+from finesse.settings import settings
+
+
+class DeviceParametersWidget(QWidget):
+    """A widget containing controls for setting a device's parameters."""
+
+    def __init__(self, device_type: DeviceTypeInfo) -> None:
+        """Create a new DeviceParametersWidget.
+
+        Args:
+            device_type: The device type whose parameters will be used
+        """
+        super().__init__()
+
+        self.device_type = device_type
+        """This value is not used within the class, but is stored for convenience."""
+
+        layout = QHBoxLayout()
+        layout.setContentsMargins(0, 0, 0, 0)
+        self.setLayout(layout)
+
+        # Make a combo box for each parameter
+        self._combos: dict[str, QComboBox] = {}
+        for param in device_type.parameters:
+            combo = QComboBox()
+            combo.setSizePolicy(QSizePolicy.Policy.Fixed, QSizePolicy.Policy.Fixed)
+
+            # Keep the "real" value along with its string representation, so that we can
+            # pass it back to the backend on device open
+            for value in param.possible_values:
+                combo.addItem(str(value), value)
+
+            if param.default_value is not None:
+                combo.setCurrentIndex(param.possible_values.index(param.default_value))
+
+            layout.addWidget(combo)
+            self._combos[param.name] = combo
+
+        # If there are saved parameter values, load them now
+        self.load_saved_parameter_values()
+
+    def set_parameter_value(self, param: str, value: Any) -> None:
+        """Set the relevant combo box's parameter value."""
+        self._combos[param].setCurrentText(str(value))
+
+    def load_saved_parameter_values(self) -> None:
+        """Set the combo boxes' parameter values according to their saved values."""
+        params = cast(
+            dict[str, Any] | None,
+            settings.value(f"device/params/{self.device_type.class_name}"),
+        )
+        if not params:
+            return
+
+        for param, value in params.items():
+            try:
+                self.set_parameter_value(param, value)
+            except Exception as error:
+                logging.warn(f"Error while setting param {param}: {error!s}")
+
+    @property
+    def current_parameter_values(self) -> dict[str, Any]:
+        """Get all parameters and their current values."""
+        return {param: combo.currentData() for param, combo in self._combos.items()}
+
+
+class DeviceTypeControl(QGroupBox):
+    """A set of widgets for choosing a device and its params and connecting to it."""
+
+    def __init__(
```

Comment:
This constructor is quite long. It might make sense to break it down into smaller chunks, roughly matching your commented sections.

Reply:
Good point. As I've since modified this code on newer branches, I'll open an issue for it and fix it later.